embassy-imxrt/src/iopctl.rs

```diff
@@ -0,0 +1,715 @@
+    ///
+    /// # Safety
+    ///
+    /// The caller MUST ensure valid port and pin numbers are provided,
+    /// and that multiple instances of [`AnyPin`] with the same port
+    /// and pin combination are not being used simultaneously.
+    ///
+    /// Failure to uphold these requirements will result in undefined behavior.
+    ///
+    /// See Table 297 in reference manual for a list of valid
+    /// pin and port number combinations.
+    #[must_use]
+    pub unsafe fn new(port: u8, pin: u8) -> Self {
+        // Calculates the offset from the beginning of the IOPCTL register block
+        // address to the register address representing the pin.
+        //
+        // See Table 297 in reference manual for how this offset is calculated.
+        let offset = ((port as usize) << 7) + ((pin as usize) << 2);
+
+        // SAFETY: This is safe assuming the caller of this function satisfies the safety requirements above.
+        let reg = unsafe { &*Iopctl::ptr().byte_offset(offset as isize).cast() };
+        Self {
+            pin_port: port * 32 + pin,
+            reg,
+        }
+    }
+
+    /// Returns the pin's port and pin combination.
+    #[must_use]
+    pub fn pin_port(&self) -> usize {
+        self.pin_port as usize
+    }
+}
+
+/// Represents a FC15 pin peripheral created at run-time from given pin number.
+pub struct FC15Pin {
+    reg: &'static PioM_N,
+}
+
+impl FC15Pin {
+    /// Creates an FC15 pin from raw pin number which can then be configured.
+    ///
+    /// This should ONLY be called when there is no other choice
+    /// (e.g. from a type-erased GPIO pin).
+    ///
+    /// Otherwise, pin peripherals should be configured directly.
+    ///
+    /// # Safety
+    ///
+    /// The caller MUST ensure valid port and pin numbers are provided,
+    /// and that multiple instances of [`AnyPin`] with the same port
+    /// and pin combination are not being used simultaneously.
+    ///
+    /// Failure to uphold these requirements will result in undefined behavior.
+    ///
+    /// See Table 297 in reference manual for a list of valid
+    /// pin and port number combinations.
+    #[must_use]
+    pub unsafe fn new(pin: u8) -> Self {
+        // Table 297:  FC15_I2C_SCL offset = 0x400, FC15_I2C_SCL offset = 0x404
+        let iopctl = unsafe { crate::pac::Iopctl::steal() };
+
+        let reg = if pin == 0 {
+            &*iopctl.fc15_i2c_scl().as_ptr().cast()
+        } else {
+            &*iopctl.fc15_i2c_sda().as_ptr().cast()
+        };
+
+        Self { reg }
+    }
+}
+
+// This allows AnyPin/FC15Pin to be used in HAL constructors that require types
+// which impl Peripheral. Used primarily by GPIO HAL to convert type-erased
+// GPIO pins back into an Output or Input pin specifically.
+embassy_hal_internal::impl_peripheral!(AnyPin);
+
+impl SealedPin for AnyPin {}
+
+embassy_hal_internal::impl_peripheral!(FC15Pin);
+
+impl SealedPin for FC15Pin {}
+
+macro_rules! impl_iopctlpin {
```

Comment:
Like the issue mentioned with enabling interrupts, the modify calls should be guarded by a critical section.